src/lib/coda_base/pending_coinbase.ml

```diff
@@ -0,0 +1,552 @@
+      (Fold.string_triples "Pending coinbases merkle tree")
+    |> of_hash
+
+  let of_digest = Fn.compose Fn.id of_hash
+end
+
+module T = struct
+  module Stack = struct
+    include Coinbase_stack.Stack
+
+    let hash (t : t) = Hash.of_digest (digest t :> field)
+  end
+
+  module Merkle_tree =
+    Sparse_ledger_lib.Sparse_ledger.Make (Hash) (Stack_pos) (Stack)
+
+  module Checked = struct
+    open Coinbase_stack
+
+    type var = Hash.var
+
+    module Merkle_tree =
+      Snarky.Merkle_tree.Checked
+        (Tick)
+        (struct
+          type value = Pedersen.Checked.Digest.t
+
+          type var = Pedersen.Checked.Digest.var
+
+          let typ = Pedersen.Checked.Digest.typ
+
+          let hash ~height h1 h2 =
+            let to_triples (bs : Pedersen.Checked.Digest.Unpacked.var) =
+              Bitstring_lib.Bitstring.pad_to_triple_list
+                ~default:Boolean.false_
+                (bs :> Boolean.var list)
+            in
+            let%bind h1 = Pedersen.Checked.Digest.choose_preimage h1
+            and h2 = Pedersen.Checked.Digest.choose_preimage h2 in
+            Pedersen.Checked.digest_triples
+              ~init:Hash_prefix.coinbase_merkle_tree.(height)
+              (to_triples h1 @ to_triples h2)
+
+          let assert_equal h1 h2 = Field.Checked.Assert.equal h1 h2
+
+          let if_ = Field.Checked.if_
+        end)
+        (struct
+          include Stack
+
+          type value = t [@@deriving sexp]
+
+          let hash = Checked.digest
+        end)
+
+    let depth = coinbase_tree_depth
+
+    type path = Pedersen.Digest.t list
+
+    type index = int
+
+    type _ Request.t +=
+      | Coinbase_stack_path : index -> path Request.t
+      | Get_coinbase_stack : index -> (Stack.t * path) Request.t
+      | Set_coinbase_stack : index * Stack.t -> unit Request.t
+      | Find_index_of_newest_stack : bool -> index Request.t
+      | Find_index_of_oldest_stack : index Request.t
+      | Reset : unit Request.t
+
+    let reraise_merkle_requests (With {request; respond}) =
+      match request with
+      | Merkle_tree.Get_path addr ->
+          respond (Delegate (Coinbase_stack_path addr))
+      | Merkle_tree.Set (addr, stack) ->
+          respond (Delegate (Set_coinbase_stack (addr, stack)))
+      | Merkle_tree.Get_element addr ->
+          respond (Delegate (Get_coinbase_stack addr))
+      | _ -> unhandled
+
+    let get t addr =
+      handle
+        (Merkle_tree.get_req ~depth (Hash.var_to_hash_packed t) addr)
+        reraise_merkle_requests
+
+    let reset () = perform As_prover.(return Reset)
+
+    let update_stack' t ~is_new_stack stack =
+      let%bind addr =
+        request_witness Stack_pos.Unpacked.typ
+          As_prover.(
+            map (read Boolean.typ is_new_stack) ~f:(fun s ->
+                Find_index_of_newest_stack s ))
+      in
+      handle
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun _s -> return stack ))
```

Comment:
isn't here where you're suppose to push the new coinbase?

Reply:
As discussed, coinbase is pushed to the stack in transaction snark. Here, the old stack is replaced with the new stack. I've now added a check for the old stack